### DISCUSSION TITLE
Rye and uv

### DISCUSSION THREAD

**Discussion by charliermarsh** [astral-sh/rye#1342] [General]
👋 Since we took over stewardship of Rye in [February](https://astral.sh/blog/uv), we've been continuing to maintain the project and ship releases, but have been primarily focused on new feature development over in [uv](https://github.com/astral-sh/uv).

Rye uses uv under-the-hood, so it's also benefited from those changes (e.g., Rye's [universal mode](https://rye.astral.sh/guide/pyproject/#toolryeuniversal) is powered by uv's universal resolver).

With the release of [uv 0.3.0](https://github.com/astral-sh/uv/releases/tag/0.3.0), uv is capable of supporting more and more Rye-like workflows. For example, uv can now download Python for you, generate lockfiles via `uv lock` and `uv sync`, and so on.

[We don't plan to deprecate Rye any time soon](https://github.com/astral-sh/rye/discussions/1164#discussioncomment-9812734), but our goal has always been to maintain it while we develop uv into a viable alternative. I suspect that uv can now replace Rye for some use cases, but I'm interested in hearing from Rye users: what is uv missing that would stop you from switching over? Where should we focus, if our goal is to enable smooth transition for Rye users in the future?

**Comment by mrkbac** [comment]
The new UV features look great!
One rye feature I like is the global python shim that automatically uses the venv with the correct python version. Also prevents me from using the system python, saved me a lot of headaches.

**Reply by yacoob** [reply]
It'd be great to keep a single canonical comparison between `uv` and `rye` somewhere, so people can figure out if their workflows are supported by `uv` yet :)

**Reply by charliermarsh** [reply]
That's a good idea. There are a few things we definitely don't support yet:

- Rye's `scripts` API
- Rye's `sources` API with different index priorities
- Anything related to Rye's shims

**Reply by zanieb** [reply]
The shim can be tracked at https://github.com/astral-sh/uv/issues/6265

**Reply by kevinli1993** [reply]
Rye's `sources` API (ability to specify a company private repo for certain packages) is needed for our use case.  I personally also enjoy Rye's shims, and think it's a major advantage over having to use, e.g., `poetry shell`.

**Reply by patrick-egenlauf** [reply]
@charliermarsh do you plan to include something like Rye's `scripts` API?

**Reply by charliermarsh** [reply]
We do yes: https://github.com/astral-sh/uv/issues/5903

**Comment by ilyagr** [comment]
**Update:** Since `uv 0.4`, it supports virtual projects.

----------

Does `uv` support [virtual projects](https://rye.astral.sh/guide/virtual/)?

As best I can tell from [the CLI reference](https://docs.astral.sh/uv/reference/cli/#uv-sync), the answer is no, but it's not completely clear. The relevant passage I found after quite a bit of searching is:

> uv will search for a project in the current directory or any parent directory. If a project cannot be found, uv will exit with an error.

Relatedly, I feel like your docs need some adjustment after version 0.3 and to make them easier to navigate (when looking for this kind of info, at least). The docs on project management ([project concepts][1], [projects guide][2]) are not trivial to find, and it's hard to guess before reading them which of them (if either) will have the needed info. After reading them, they do not seem to address the question (I could have missed the answer hiding somewhere, though; I also haven't experimented with it yet and things might become clearer once I do).

[1]: https://docs.astral.sh/uv/concepts/projects/
[2]: https://docs.astral.sh/uv/guides/projects/#pyprojecttoml

Rye is a good example to follow for clear docs. It is easier for `rye` to achieve that, since the tool has a more focused scope, but I'm sure that this is doable for `uv` too. I guess this will be my last bit of feedback for now: `uv` can do a lot of cool things, but I think the UI and docs could do a better job of making it easy to discover just the right set of tools for each of the use-cases (e.g. "set up a `pyproject.toml` to manage the MkDocs dependencies for building docs for a non-python project, as well as a workflow to make it easy for Rust developers to run MkDocs", in my case).

**Reply by zanieb** [reply]
Thanks for the feedback! We'll definitely need to fill out some of the project concept documentation, like "what is the definition of a project" and "how to use virtual projects". It's hard to go from zero to one hundred percent coverage in one go. What made them hard to find?

In short, I believe we don't support virtual projects — except at the root of a workspace, in which case only development dependencies are supported. I think we're going to expand support for this in the near future though.

**Reply by ilyagr** [reply]
>  I believe we don't support virtual projects — except at the root of a workspace, in which case only development dependencies are supported.

That is exactly what I would need, if I understood you correctly. Are there docs on how to set that up?

**Reply by zanieb** [reply]
That's discussed at https://docs.astral.sh/uv/concepts/workspaces/#workspace-roots

**Reply by ilyagr** [reply]
Thanks, I'll try it out! I'll let you know if I have trouble when I do, but it sounds very promising.

> What made them hard to find?

I can speculate a bit. I'm less certain about the reason than I am about the fact that they were difficult to find without your help.

For one, I don't need any other features "workspaces" provide, so I never looked there.

For another, while both docs have a lot of useful information, I don't think the split between the "project concepts" and the "project guide" in the docs is working the way (I'd guess) it was supposed to. Neither of these docs is easy to navigate, as a guide or as a reference. Fixing this wouldn't be a trivial task, but I think they should be reorganized once the features from `uv 0.3` are more or less stable and the workflow that the users should follow is clear.

**Reply by ilyagr** [reply]
It doesn't actually seems to be working, at least not immediately. I'm not sure whether this is a a fundamental limitation, a bug, or maybe all I need is a hint of what my config is missing. I'm guessing, it's a case of "not quite implemented yet" :).

Just in case it's unexpected, here's what I did.

I did `uv init --virtual` and then populated the dev-dependencies as follows:

```toml
[tool.uv.workspace]
members = []

[tool.uv]
dev-dependencies = [
    "mkdocs>=1.6, <1.7",
    "mkdocs-material>=9.5.24, <9.6",
    # Versioning of documentation
    "mike>=2.1.1, <3",
    # (Py)Markdown extensions
    "mdx-truly-sane-lists >=1.3, <2",
    "mdx-breakless-lists >=1.0.1, <1.1",
    # Allows setting up redirects when renaming docs files
    "mkdocs-redirects >=1.2.1, <1.3",
    # Used for the CLI reference
    "mkdocs-include-markdown-plugin >=6.0.4, <6.1",
]
```

Now, I get:

```console
$ RUST_BACKTRACE=1 uv run mkdocs
thread 'main' panicked at crates/uv-resolver/src/lock.rs:455:22:
could not find root
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
$ uv --version
uv 0.3.0 (Homebrew 2024-08-20)
```

For reference, the Rye config I'm trying to emulate is:

```toml
[project]
name = "jj-docs"
dependencies = []
requires-python = ">= 3.8"

[tool.rye]
managed = false
virtual = true
universal=true
generate-hashes=true
dev-dependencies = [
    "mkdocs>=1.6, <1.7",
    "mkdocs-material>=9.5.24, <9.6",
    # Versioning of documentation
    "mike>=2.1.1, <3",
    # (Py)Markdown extensions
    "mdx-truly-sane-lists >=1.3, <2",
    "mdx-breakless-lists >=1.0.1, <1.1",
    # Allows setting up redirects when renaming docs files
    "mkdocs-redirects >=1.2.1, <1.3",
    # Used for the CLI reference
    "mkdocs-include-markdown-plugin >=6.0.4, <6.1",
]
```

It doesn't really matter, but [here's the Poetry config we are currently using](https://github.com/martinvonz/jj/blob/92403d7d11eb8e94da61e5a1f5237d424c052b32/pyproject.toml). I'm trying to decide whether to switch to Rye or UV (so that our contributors don't need to install Python+pipx+poetry manually if they don't generally use Python), and when...

**Reply by zanieb** [reply]
A panic! Sounds like a bug. We'll look into that immediately. Thanks!

**Reply by charliermarsh** [reply]
(Thanks, I'll take a look now.)

**Reply by charliermarsh** [reply]
Unfortunately I haven't been able to reproduce this. I copied that `pyproject.toml` into a new directory and ran:

```
❯ uv run mkdocs
Using Python 3.12.1
Creating virtualenv at: .venv
Installed 41 packages in 294ms
Usage: mkdocs [OPTIONS] COMMAND [ARGS]...

  MkDocs - Project documentation with Markdown.

Options:
  -V, --version         Show the version and exit.
  -q, --quiet           Silence warnings
  -v, --verbose         Enable verbose output
  --color / --no-color  Force enable or disable color and wrapping for the output. Default is auto-detect.
  -h, --help            Show this message and exit.

Commands:
  build      Build the MkDocs documentation.
  get-deps   Show required PyPI packages inferred from plugins in mkdocs.yml.
  gh-deploy  Deploy your documentation to GitHub Pages.
  new        Create a new MkDocs project.
  serve      Run the builtin development server.
```

Can you think of anything else that might differ?

**Reply by charliermarsh** [reply]
Oh I managed to reproduce with a slightly different workflow. Will fix, thanks.

**Reply by ilyagr** [reply]
> Can you think of anything else that might differ?

Yes. Somehow, I ended up with the following `uv.lock`:

```toml
version = 1
requires-python = ">=3.12"
```

An dir with that uv.lock and my pyproject.toml (just 2 files) reproduces the bug for me.

Meanwhile, removing the `uv.lock` in my actual project seems to cause `mkdocs` to start working just like it should! 🎉 It generates a working `uv.lock` as well.

**Reply by charliermarsh** [reply]
Perfect, thank you!

**Reply by charliermarsh** [reply]
Fixed: https://github.com/astral-sh/uv/pull/6291

**Reply by quodlibetor** [reply]
For people coming to this discussion now, uv 0.4 made significant improvements and changes to the virtual package/no-code package feature set: https://github.com/astral-sh/uv/releases/tag/0.4.0

**Reply by dynamicalsystem** [reply]
Still continuing my migration to `uv` from `rye`.  I'm trying to use `uv init --virtual` but the `--virtual` option doesn't seem to do anything, there's no mention of it as an option in either the webpage docs or the man page and the folder it produces has a `hello.py` file which seems redundant.  I can see that [6720](https://github.com/astral-sh/uv/pull/6720/files) removes the concept of virtual workspaces from the virtual|explicit workspace roots explainer and refers to it as legacy... and even the [Workspace Roots](https://github.com/astral-sh/rye/discussions/1342#discussioncomment-10400642) section mentioned above has been excised.

Has `--virtual` been deprecated or am I doing it wrong?

**Reply by charliermarsh** [reply]
Yeah we removed the virtual concept in favor of “projects that aren’t packages”. The app flag to uv init will create a project that uv will not treat as a package, so it won’t be built and installed into the environment, like Rye’s virtual workspace root concept.

**Comment by jessekrubin** [comment]
FWIW: the most alarmingly great thing about rye is that it works just as good on windows (many python things don't work good on windows). My place of work develops for rhel and windows, so having both "work good" is crucial.

**Comment by shamrin** [comment]
I am a bit worried that you plan to make people switch from rye to uv eventually. The confusion has some potential to end both projects. (Perl 6 as the most egregious example.)

Have you considered for the future to simply rename `uv` to `rye` and make the most important sub commands compatible?

**Reply by mitsuhiko** [reply]
@shamrin my original goal was (unless a better plan emerges) to have a "final" release of `rye` eventually that just dispatches to uv under the hood and upgrades the project if necessary to uv tool conventions to make the upgrade transparent.

**Comment by gwdekker** [comment]
Probably a bit lower on your list, but we use the polylith plugin for rye to make our monorepo setup work: https://github.com/DavidVujic/python-polylith-example-rye

apart from the nice tooling the polylith CLI offers, key functionality and setup here is:
- Allow for relative paths (as far as I can see uv already can do that)
- Allow for a setup where there is one root pyproject.toml and also pyproject.tomls in the /projects section for deployable units, and one lock file at the root of the repo.

I am not a rust programmer, so maybe this is all already covered by the workspaces concept?

**Reply by DavidVujic** [reply]
I'm happy to hear about the Polylith usage! I have begun to look into the workspaces feature of uv (and I guess also Rye), and how it can be combined with the Polylith tooling.

**Reply by amanmibra** [reply]
following here @DavidVujic, we also use polylith and appreciate the work you put in to make it easier to use.

**Comment by bluss** [comment]
I think `rye init --script` is a minor one that would be nice to have in uv, because it shows by example how to set up a project with executables.

Edit: Uv now does this for uv init --app --package

**Reply by my1e5** [reply]
+1

As it currently stands, starting from
```
uv init myproject
```
All you technically need to do to make your project executable is to add
```
[project.scripts]
myproject = "myproject:hello"
```
to the `pyproject.toml` file. Then
```
$ uv run myproject
Hello from myproject!
```
But would be nice if there was a `--script` flag that set this up and did whatever the best practice is in terms of layout. I know that Rye also creates a `__main__.py` file and uses a `main` function in `__init__.py`  instead of `hello`.

**Comment by dream-dasher** [comment]
Rye's support for [Rust modules out-of-the-box](https://rye.astral.sh/guide/rust/) is quite nice.  It was something that I'd love to see get even more love.  Any chance uv will focus on supporting this as well.  It enhances both languages considerably I feel, if that friction is removed in what's likely a future standard too.

I'd like to be able to have a python directory with notebooks (a la [Zed repl](https://zed.dev/docs/repl)) inside my rust repos and be able to setup rust in my python repos to add code easily.  And I think reducing that friction would bring a lot to the table.


(Nice work on supporting single-file scripts, btw.  I'm looking forward to that coming out of nightly in Rust for the 2024 version.)

**Reply by vivienm** [reply]
> Rye's support for [Rust modules out-of-the-box](https://rye.astral.sh/guide/rust/) is quite nice. It was something that I'd love to see get even more love. Any chance uv will focus on supporting this as well. It enhances both languages considerably I feel, if that friction is removed in what's likely a future standard too.

This seems to work by adding the following lines to the pyproject.toml file (as rye does):

```toml
[build-system]
requires = ["maturin>=1.7.0"]
build-backend = "maturin"

[tool.maturin]
python-source = "python"
module-name = "my_project._lowlevel"
features = ["pyo3/extension-module"]
```

**Reply by bluss** [reply]
A maturin template is now implemented in uv and will be in the next release.

**Comment by jgeralnik** [comment]
I just migrated from rye to uv. It was relatively straightforward - got most of the way there by replacing [tool.rye] with [tool.uv] in pyproject.toml

The docker build required slightly more changes because previously I was copying in rye's `requirements.lock` and installing it with regular pip without needing rye in the build process at all. The [docker guide](https://docs.astral.sh/uv/guides/integration/docker/) (once I found it) was straightforward to follow. I did discover that I could just use uv compile and continue using raw `requirement.txt` files in my docker image but uv is so much faster that I left it that way anyways.

Likewise the github action was straightforward to set up by following [the guide](https://docs.astral.sh/uv/guides/integration/github/#installation) as well.

**Reply by jgeralnik** [reply]
One thing I would add to the Docker guide is to run `uv sync --frozen --no-dev` instead of just `uv sync`

**Reply by nickchomey** [reply]
Is there any doc available for migrating from rye to uv?

**Comment by uwu-420** [comment]
Really excited to make the switch to uv! Thanks for all the great work :)

If I understood the docs about [dependencies](https://docs.astral.sh/uv/concepts/dependencies/) and [alternative indexes](https://docs.astral.sh/uv/guides/integration/alternative-indexes/) correctly, I can only specify the source for a specific dependency and I could set `UV_EXTRA_INDEX_URL` but there is no way to add additional indexes like I can do with rye? See [here](https://rye.astral.sh/guide/faq/#how-do-i-install-pytorch)

```toml
[[tool.rye.sources]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cpu"
```

I have a bunch of dependencies from my own index and it would be cool to just include it as an additional source like I can do with rye. Or are there some bad consequences when doing that? Would be interested if someone else would like this feature.

**Reply by charliermarsh** [reply]
Yeah that's correct -- we only support the pip-like `--extra-index-url` API. We do want to add an index priority API but we cut it from scope for the release -- hopefully will come soon.

**Reply by Systemcluster** [reply]
This is a very important feature for anything related to GPU compute (language models etc.) where libraries like `pytorch` provide different sources for different CUDA versions including subdependencies.

**Reply by JonZeolla** [reply]
I'm in the same boat and this would be quite valuable to me. Similarly we are looking forward to index pinning

**Comment by ciscorucinski** [comment]
On Linux, the thing that has stopped me the most from updating Python has been the worry of accidentally overwriting my system Python. I want to install every major point release, and keep them updated. I also would potentially like to try out new beta versions if needed, but I want all of these installs to be easily discovered within PyCharm, even with new projects.

It seems like Rye might allow this (or a subset) via `rye toolchain ...`, right? I just heard about this Rye feature earlier today. But is this also possible (or extended) under uv?

**Reply by bluss** [reply]
See the `uv python` subcommand, it's basically the same feature

**Reply by grazingbunny** [reply]
I haven't migrated to using `uv` yet, however I have also been a user of `rye toolchain` to manage my Python installations.

It looks like `uv` [supports this use case](https://docs.astral.sh/uv/guides/install-python/) via `uv python`.

Currently the only think missing is something that works like the `rye` shim, but [it is planned](https://github.com/astral-sh/uv/issues/6265).

**Reply by ciscorucinski** [reply]
> Currently the only think missing is something that works like the `rye` shim, but [it is planned](https://github.com/astral-sh/uv/issues/6265).

This is what is preventing it from becoming globally available to PyCharm? Basically this shim would be just like downloading Python from the official source and installing it in the correct directory (to not override system python)?

**Reply by scottyeager** [reply]
It's worth noting that both Rye and uv install portable builds that deviate somewhat from what you'd get from a distro package manager or building from source (such as via pyenv). You can read about those builds and their quirks [here](https://gregoryszorc.com/docs/python-build-standalone/main/).

The shim feature enables automatic and transparent use of different versions of Python in different shell contexts. That can be per folder and/or globally. Shims only affect users who enable them, so anything root or other users do with system Python will be unaffected.

Pycharm has its own feature for using different Python versions with different projects. That should work with any solution for getting different Python builds, be it uv, Rye, pyenv, or something else. You just need to know the paths.

**Comment by asmith26** [comment]
Does `uv` support https://github.com/dhatim/python-license-check?

In particular, because `rye` outputs standard *requirements.txt* files, this lib works well with a `rye` project. Thanks!

**Reply by asmith26** [reply]
Looks like this has just been added, many thanks! https://x.com/charliermarsh/status/1829620595048468480

**Comment by varioustoxins** [comment]
Build and publish? Or have I missed something?

**Reply by zanieb** [reply]
Definitely on the short-term roadmap. Documented as not supported at https://docs.astral.sh/uv/guides/publish/

We're tracking at https://github.com/astral-sh/uv/issues/6278 / https://github.com/astral-sh/uv/issues/1510

**Comment by boenshao** [comment]
Would uv support commands like `fmt` and `test` in the future? Or will it focus on managing the Python environment?

**Reply by my1e5** [reply]
(not a direct answer to your question, but for those worried about losing `fmt` and `test` you can still get this kind of functionality by adding `ruff` and `pytest` as dev dependencies like `uv add --dev ruff`. Then you can add a task runner library (see https://github.com/astral-sh/uv/issues/5903) and assign a key like `fmt = "ruff format ."` in your `pyproject.toml` file.)

(Or you could install `ruff` as a global tool. But I like keeping the `ruff` version and dependency contained within the `pyproject.toml` on a per project basis. In fact, I would do this for my Rye managed projects as well. The built-in Rye ruff version was always a bit behind the latest release for example)

**Reply by lake-hope** [reply]
I really want `uv` to ship with support for `uv fmt`, `uv test`, and `uv lint` out of the box, and use Black, pytest, and Ruff for those, respectively — just like Rye. One of the best things about Rust and Cargo is that they standardize the tooling for formatting, linting, and testing the project, so people don't have to pick from a myriad of different formatters, linters, and test frameworks just to get started. Python has had too many options for each of those tools, and I don't think that forcing choice is good here — I've worked on projects where people spent *way* too much time figuring out (and sometimes arguing) which one to use. And even if Black and pytest have become the de-facto standard (and I hope Ruff gets there soon, too), having to manually install and configure them with `uv` — and then needing a longer command to run them — adds extra steps and increases cognitive load. If people want to use something else, like pylint, or unittest, then that can be supported as a custom subcommand. But I think the Python community *really* needs good default tooling that doesn't require extra steps or configuration to use. 

In Rust, I can open a project, and run `cargo fmt` or `cargo test` or `cargo clippy`, and in almost all projects, it will automagically know what to do and where to find my code. Rye provides that, too. But `uv` currently doesn't. I want to promote `uv` as a definitive *project* management tool, and a solution to the (honestly often soul-crushing) work of managing Python projects, so that someone can do `uv init` and get everything they need, and not have all the enormous mental overheads we have with existing tooling that isn't Rye.

**Reply by ryan-minato** [reply]
From my perspective, `fmt`, `test`, and `lint` should be part of the default template of uv, but it should also allow users the flexibility to replace them if needed. While it's fine to provide default settings, forcing a specific tech stack on everyone isn't ideal.

Currently, `rye` defaults to using `pytest`, but it doesn't create a `tests` folder. A better practice would be to automatically add pytest to the dev-dependencies section and provide a shortcut under something like `tool.rye.scripts` (well, custom commands are on UV's TODO list), while also creating a default `tests` folder. 
If someone must using `unittest`, they could simply adjust the command in `tool.rye.scripts` instead of having a redundant top-level subcommand that they'll never use.

If compatibility must be considered, perhaps `uv test` could be mapped to `uv run test`.

btw, I completely understand the desire to escape the never-ending decisions around toolchains (especially since I sometimes have to do front-end work, and the front-end toolchain is a real nightmare). However, in my case, mandating the use of ruff hasn't actually eased the debates within my team. We still argue over whether boolean parameters should be allowed, whether it's okay to write error messages directly inside parentheses, and whether assert statements should be permitted.

**Reply by dynamicalsystem** [reply]
Plus one for `uv test --package <package>`. I have started to migrate `rye` functions to `uv` and this is the immediate sticking point in my makefile.

**Comment by sena-nana** [comment]
Will uv support rye.virtual configuration? I'm not sure there is a configuration that will accomplish the same thing. I don't want to install the project itself into the environment.

**Reply by baggiponte** [reply]
I think you can since `uv` 0.4.0! https://docs.astral.sh/uv/concepts/projects/#creating-projects

**Comment by SamoylovMD** [comment]
I personally use Rye to build wheels and publish them to pypi (`rye build` and `rye publish` correspondingly). I don't see such functionality in uv, does that mean you're planning to leave building and publishing for external tools like hatch? In this case this will go against Rye purpose as being the single entrypoint for managing Python projects as we'll need again this multiverse of tools instead of one single solution.

**Reply by dflock** [reply]
They are planning to support both `uv build` and `uv publish`: https://github.com/astral-sh/rye/discussions/1342#discussioncomment-10413566

**Comment by AucaCoyan** [comment]
A low hanging fruit is that `uv init` doesn't create a git repository (I couldn't find it on the `uv init --help`), but `rye init` does.
It's not stopping me at all, but would be a great feature to have if we want to have cargo in python 🥰

**Reply by charliermarsh** [reply]
There's an open PR for this thankfully: https://github.com/astral-sh/uv/pull/5476

**Comment by benrutter** [comment]
For me, aside from liking the python shim (which somebody else has mentioned) the main thing stopping me from swapping to uv (from rye) is for team projects where rye is working well, the different locations for dev dependencies in `pyproject.toml` mean I'd need to convince the rest of the team to swap from rye to uv too, which isn't the easiest argument when rye is currently working great for everyone.

It's kinda caused by python not having a standard dev requirements protocol, but it would be amazing for some kind of magic setting to tell uv "no actually look at this section for dev requirements instead".

**Reply by benrutter** [reply]
(even better would be if the python community standardised on a way of capturing dev requirements, but that's probably a lot harder to achieve as a single project than a settings option 😅)

**Reply by harkabeeparolus** [reply]
> (even better would be if the python community standardised on a way of capturing dev requirements, but that's probably a lot harder to achieve as a single project than a settings option 😅)

Guess what, they did, and it's done! 👍🏻 [PEP 735 – Dependency Groups in pyproject.toml](https://peps.python.org/pep-0735/)

**Comment by AnirudhG07** [comment]
I just found out rye and uv, and from the conversation, it seems rye and uv have both its pros and cons, and uv being under development, may potentially replace rye(?)... So should i start using uv directly? I only need for basic python packages, not yet building large scale projects...

**Comment by phi-friday** [comment]
If only `build`, `publish` were added, I would change it to `uv`.
It would be nice if the `tool.rye.virtual` option was also available, but it's not required.

**Reply by phi-friday** [reply]
I tried migrating from `rye` to `uv`.
I needed the following features more than I expected

1. `tool.rye.scripts`
2. `rye version <some version>`

**Reply by my1e5** [reply]
`tool.rye.scripts`-like functionality can be obtained using 3rd party libraries - see https://github.com/astral-sh/uv/issues/5903

**Reply by phi-friday** [reply]
One of the last features I needed was to generate a `requirements-dev.txt` or `requirements-dev.lock` containing `tool.uv.dev-dependencies`.

I understand that this is not yet supported.
https://github.com/astral-sh/uv/issues/6429
https://github.com/astral-sh/uv/issues/6007

**Reply by bluss** [reply]
build and publish are available in uv now :slightly_smiling_face:

**Reply by bluss** [reply]
And uv export can generate requirements files -- with dev, without dev, only dev, etc.

**Comment by nickchomey** [comment]
Is there a doc for migrating from rye to uv? It should be a relatively simple process, but would still be good to have instructions

**Reply by jgeralnik** [reply]
No doc, but really - just place dev-dependencies under [tool.uv] instead of [tool.rye] (didn't need any other settings from there) and then run `uv sync`.
The more complex part if if you're doing something else with your project (running in docker, publishing to pypi, etc). If you have any specific questions I'd be happy to help

**Reply by lucianosrp** [reply]
Also, if that helps, I wrote a [simple tool](https://github.com/lucianosrp/rye-uv) to migrate from rye to uv and I am planning to keep track on the different changes. 
For now, it only  converts to uv~0.3, but I will be adding the conversion of `tool.rye.virtual`  relatively soon.

**Reply by mandarvaze** [reply]
@jgeralnik I replaced `tool.rye` with `tool.uv` in my `pyproject.toml` and executed `uv sync`
I got `unknown field `virtual`` - because I have the following entry 

```toml
[tool.rye]
managed = true
virtual = true
dev-dependencies = [
    "ruff>=0.5.5",
    "pytest>=8.3.2",
]
```

Any suggestions ?

**Reply by lucianosrp** [reply]
@mandarvaze, you should be able to replace `virtual=true` with `package=false` 
I made an "unofficial" conversion table between rye and uv [here](https://github.com/lucianosrp/rye-uv?tab=readme-ov-file#conversion)

**Reply by mandarvaze** [reply]
@lucianosrp Thanks. The error went away after making changes suggested by you.

**Comment by Super1Windcloud** [comment]
I  think  that uv  compared with rye, uv's  function and   feature  is more simple and practical

**Comment by ryan-minato** [comment]
The feature of generating the project layout for pyo3 is very important to me—it was almost the sole reason I switched from poetry to rye.  Some of my work-related projects are highly performance-sensitive, and pyo3 provides significant benefits. However, before using rye, I couldn't find a tool that allowed me to initialize a project in under 30 seconds.

By the way, I’ve been using uv to manage my experiments because its locking mechanism is simple, intuitive, and fast.If uv had better support for pyo3, I would have switched to uv without hesitation.

**Reply by harkabeeparolus** [reply]
This is available now: `uv init --build-backend maturin --app foo`

**Comment by 0x00cl** [comment]
I think that uv and rye docs should point to some of their differences and overall goal. I was a bit confused on why they were developing two projects, and which one should I use, specially because the homepage of each tools describe them as the same thing.

rye:
> Rye is a comprehensive project and package management solution for Python

uv:
> An extremely fast Python package and project manager, written in Rust.

The only answer is from astral blog which I doubt developers will read or even find.

> We'll maintain [Rye](https://github.com/mitsuhiko/rye) as we expand [uv](https://github.com/astral-sh/uv) into a unified successor project, to fulfill our [shared vision](https://rye-up.com/philosophy/) for Python packaging.

Letting developers know uv is the intended successor should cause less problems in the future.

EDIT: I want to clarify that the docs should contain that information and not hidden behind blogs.

**Comment by impredicative** [comment]
> Letting developers know uv is the intended successor

If rye becomes a progressively thin layer to uv while maintaining its functionality, that won't be so bad either.

**Reply by chaneyzorn** [reply]
> If rye becomes a progressively thin layer to uv while maintaining its functionality, that won't be so bad either.

If Maintainer asked this:

> what is uv missing that would stop you from switching over?

Why not just announce `uv is the intended successor`?

Speaking clearly and firmly can avoid many problems.

**Reply by zanieb** [reply]
It is the intended successor. I think this has been fairly clearly stated. e.g. from Armin https://lucumr.pocoo.org/2024/8/21/harvest-season/

**Comment by Kristina-Pianykh** [comment]
The great feature I thought rye has is the string interpolation for index URL in `pyproject.toml`. I see this issue has been raised and tracked in `uv`: https://github.com/astral-sh/uv/issues/5734
I'm setting up a project for an internal library for my team and passing the argument for the alternative index with the whole link seems cumbersome to me as opposed to storing it in `pyproject.toml`. Every time I need to add or remove a package from deps or perform a sync, I have to provide a long index url and make sure to paste my credentials correctly so this is just jarring in usage. Until uv implements this feature, I will have to push for using rye.

**Reply by JockeTF** [reply]
Hard-coding index URLs and setting the  `UV_INDEX_<NAME>_PASSWORD` environment variables covered this use-case for me.

- https://docs.astral.sh/uv/configuration/indexes/

**Comment by hemna** [comment]
Why would anyone use rye over uv?  It's entirely unclear why I would ever use rye, when I use uv.

**Reply by astrojuanlu** [reply]
This thread is 1.5 years old, rye is now no longer developed https://github.com/astral-sh/rye/pull/1476 

Maybe Astral could archive the repo now
